src/toolchain/toolchain.ts

```diff
@@ -0,0 +1,123 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the VSCode Swift open source project
+//
+// Copyright (c) 2021 the VSCode Swift project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of VSCode Swift project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import * as fs from "fs/promises";
+import * as path from "path";
+import * as plist from "plist";
+import configuration from "../configuration";
+import { execFile, execSwift } from "../utilities/utilities";
+import { Version } from "../utilities/version";
+
+/**
+ * Contents of **Info.plist** on Windows.
+ */
+interface InfoPlist {
+    DefaultProperties: {
+        XCTEST_VERSION: string | undefined;
+    };
+}
+
+export class SwiftToolchain {
+    constructor(
+        public swiftVersionString: string,
+        public swiftVersion: Version,
+        public toolchainPath?: string,
+        public developerDir?: string
+    ) {}
+
+    static async create(): Promise<SwiftToolchain> {
+        const version = await this.getSwiftVersion();
+        const toolchainPath = await this.getToolchainPath();
+        const developerDir = await this.getDeveloperDir();
+        return new SwiftToolchain(version.name, version.version, toolchainPath, developerDir);
+    }
+
+    /**
+     * @returns path to Toolchain folder
+     */
+    private static async getToolchainPath(): Promise<string | undefined> {
+        if (configuration.path !== "") {
+            return path.dirname(path.dirname(configuration.path));
+        } else if (process.platform === "darwin") {
+            const { stdout } = await execFile("xcrun", ["--find", "swiftc"]);
+            const swiftc = stdout.trimEnd();
+            return path.dirname(path.dirname(path.dirname(swiftc)));
+        } else if (process.platform === "linux") {
+            const { stdout } = await execFile("which", ["swiftc"]);
+            const swiftc = stdout.trimEnd();
+            return path.dirname(path.dirname(path.dirname(swiftc)));
+        } else if (process.platform === "win32") {
+            const { stdout } = await execFile("where", ["swiftc"]);
+            const swiftc = stdout.trimEnd();
+            return path.dirname(path.dirname(path.dirname(swiftc)));
+        }
+        return undefined;
+    }
+
+    /**
+     * @returns path to developer folder, where we find XCTest
+     */
+    private static async getDeveloperDir(): Promise<string | undefined> {
+        try {
+            switch (process.platform) {
+                case "darwin": {
+                    const { stdout } = await execFile("xcode-select", ["-p"]);
+                    return stdout.trimEnd();
+                }
+
+                case "win32": {
+                    const developerPath = process.env.DEVELOPER_DIR;
```

Comment:
This should be exactly what `DeveloperDir` is on Windows, so I think the naming is improper. Change it back into `getXCTestDir` instead?

Reply:
Sorry yeah I got that one wrong.

Reply:
I've split it into two functions `getDeveloperDir` and `getXCTestPath`

Reply:
@stevapple would you be able to verify I haven't broken Windows